velox/exec/fuzzer/WindowFuzzer.cpp

```diff
@@ -205,6 +257,145 @@ std::vector<SortingKeyAndOrder> WindowFuzzer::generateSortingKeysAndOrders(
   return results;
 }
 
+template <typename T>
+const VectorPtr WindowFuzzer::buildKRangeColumn(
+    const VectorPtr& frameBound,
+    const std::string& columnName,
+    const VectorPtr& orderByCol,
+    const core::WindowNode::BoundType& frameBoundType,
+    const core::SortOrder& sortOrder,
+    bool isColumnBound,
+    bool isOffsetColumn) {
+  auto type = CppToType<T>::create();
+  const auto size = vectorFuzzer_.getOptions().vectorSize;
+  const SelectivityVector allRows(size);
+  DecodedVector decodedVector(*orderByCol, allRows);
+  BufferPtr values = AlignedBuffer::allocate<T>(size, pool_.get());
+  BufferPtr nulls = allocateNulls(size, pool_.get());
+  auto rawValues = values->asMutableRange<T>();
+  auto* rawNulls = nulls->asMutable<uint64_t>();
+
+  for (auto j = 0; j < size; j++) {
+    if (decodedVector.isNullAt(j)) {
+      bits::setNull(rawNulls, j, true);
+    } else {
+      const T frameBoundValue = isColumnBound
+          ? frameBound->as<FlatVector<T>>()->valueAt(j)
+          : frameBound->as<ConstantVector<T>>()->valueAt(0);
+      if (isOffsetColumn) {
+        std::optional<T> offset = genOffsetAtIdx<T>(
+            decodedVector.valueAt<T>(j),
+            frameBoundValue,
+            frameBoundType,
+            sortOrder);
+        VELOX_USER_CHECK(
+            offset.has_value(),
+            "Offset cannot be generated: orderBy key type {}, sortOrder ascending {}, frameBoundType {}",
+            type->toString(),
+            sortOrder.toString(),
+            core::WindowNode::boundTypeName(frameBoundType));
+        rawValues[j] = offset.value();
+      } else {
+        rawValues[j] = frameBoundValue;
+      }
+    }
+  }
+
+  return std::make_shared<FlatVector<T>>(
+      pool_.get(), type, nulls, size, values, std::vector<BufferPtr>{});
+}
+
+template <typename T>
+const std::string WindowFuzzer::addKRangeOffsetColumnToInputImpl(
+    std::vector<RowVectorPtr>& input,
+    FrameMetadata& frameMetadata,
+    const core::WindowNode::BoundType& frameBoundType,
+    const SortingKeyAndOrder& orderByKey,
+    const std::string& columnName,
+    const std::string& offsetColumnName) {
+  // Use columns as frame bound 50% of time.
+  bool isColumnBound = vectorFuzzer_.coinToss(0.5);
+  const auto type = CppToType<T>::create();
+  VectorPtr constantFrameBound =
+      isColumnBound ? nullptr : vectorFuzzer_.fuzzConstant(type, 1);
+  VectorPtr columnFrameBound;
+  // Generate frame bound (constant/column) without nulls.
+  VectorFuzzer::Options options = vectorFuzzer_.getOptions();
+  auto nullRatio = options.nullRatio;
+  options.nullRatio = 0;
+  vectorFuzzer_.setOptions(options);
```

Comment:
nit: You can use ScopedVarSetter here so that you don't need to reset the option later.